### PR TITLE
Watch downloads

### PR DESCRIPTION
**Pull request by Frando**
Companion PR to hyperspace-org/hyperspace#5, see there for details.

### CODE REVIEW

**Comment by mafintosh** [comment]
Nice! Instead of adding two new methods couldn’t we trigger this when a new download listener has been added?

**Comment by Frando** [comment]
@mafintosh I removed the two methods and now it just works by registering an event listener for the `download` event. Also rebased both this and the two companion PRs (in hyperspace and rpc) on top of master.

File: index.js [hunk 1]
```diff
@@ -358,6 +363,18 @@ class RemoteHypercore extends Nanoresource {
     this._extensions = new Map()
     this._onwaits = new FreeMap(1)
 
+    // Track listeners for the download event, and enable/disable download watching.
+    this.on('newListener', (event, _listener) => {
```
Comment (mafintosh) [inline index.js]: _listener is unused?

File: index.js [hunk 2]
```diff
@@ -358,6 +363,18 @@ class RemoteHypercore extends Nanoresource {
...
+      if (event === 'download' && !this.listenerCount(event)) {
+        this._watchDownloads()
+      }
+    })
+    this.on('removeListener', (event, _listener) => {
```
Comment (mafintosh) [inline index.js]: _listener is unused?

Comment (Frando) [reply index.js]: fixed

File: index.js [hunk 3]
```diff
@@ -568,6 +590,22 @@ class RemoteHypercore extends Nanoresource {
     return rsp.bytes
   }
 
+  async _watchDownloads () {
+    try {
+      if (!this.opened) await this.open()
+      if (this.closed) return
+      await this._client.hypercore.watchDownloads({ id: this._id })
```
Comment (mafintosh) [inline index.js]: watchDownloadsNoReply and no await

File: index.js [hunk 4]
```diff
@@ -568,6 +590,22 @@ class RemoteHypercore extends Nanoresource {
...
+    } catch (_) {}
+  }
+
+  async _unwatchDownloads () {
+    try {
+      if (!this.opened) await this.open()
+      if (this.closed) return
+      await this._client.hypercore.unwatchDownloads({ id: this._id })
```
Comment (mafintosh) [inline index.js]: unwatchDownloadsNoReply and no await

**Pull request merged**